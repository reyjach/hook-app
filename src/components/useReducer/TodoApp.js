import React, { useEffect, useReducer } from 'react';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


const init = () => {

    const value = JSON.parse(localStorage.getItem('todos'))

    if(value === null){
        return [];
    }
    return value;
    


    /*return [{
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false
    }]; */
}

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init );

    

    useEffect(() => {
      localStorage.setItem('todos', JSON.stringify( todos ));
    }, [todos])
    

    const handleDelete = (todoId) => {

        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = ( newTodo ) => {

        dispatch({
            type: 'add',
            payload:newTodo
        } );
    }

  return (
    <div>
        <h1>TodoApp ( { todos.length } )</h1>
        <hr />

        <div className="row">
            <div className="col-7">
                <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete}/>
            </div>

            <div className="col-5">
                <TodoAdd handleAddTodo={handleAddTodo}/>
            </div>

        </div>
    </div>
  )
}
