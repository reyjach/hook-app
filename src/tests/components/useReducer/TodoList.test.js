import { shallow } from "enzyme";
import { TodoList } from "../../../components/useReducer/TodoList";

import { demoTodo } from "../../fixture/demoTodo";


describe('pruebas en el componente TodoList', () => { 

    const handleDelete = jest.fn();

    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList 
            todos = {demoTodo}
            handleToggle = {handleToggle}
            handleDelete = {handleDelete}
        />
    )

    beforeEach ( () => { 
        jest.clearAllMocks();

        const wrapper = shallow(
            <TodoList 
                todos = {demoTodo}
                handleToggle = {handleToggle}
                handleDelete = {handleDelete}
            />
        )
    });

    test('Debe de mosrtrarse correctamamente', () => { 

        expect( wrapper ).toMatchSnapshot();


     });

     test('Debe de tener dos TodoList', () => { 

        expect( wrapper.find('TodoListItem').length ).toBe( demoTodo.length );

        expect( wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual( expect.any(Function) );

      })
 })