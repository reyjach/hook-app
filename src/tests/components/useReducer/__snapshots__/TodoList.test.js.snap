// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el componente TodoList Debe de mosrtrarse correctamamente 1`] = `
<Fragment>
  <ul
    className="list-group list-group-flush"
  >
    <TodoListItem
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      i={0}
      key="1"
      todo={
        Object {
          "desc": "Aprender react",
          "done": false,
          "id": 1,
        }
      }
    />
    <TodoListItem
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      i={1}
      key="2"
      todo={
        Object {
          "desc": "Aprender mongo",
          "done": false,
          "id": 2,
        }
      }
    />
  </ul>
</Fragment>
`;
